### ISSUE TITLE
Proposal: Add IdnMapping Span-based APIs

### ISSUE THREAD

**Issue by MihaZupan** [dotnet/runtime#32411]
The current `IdnMapping` API accepts/returns strings and throws on invalid input.

I propose a set of Span-based APIs to avoid allocations and verifying the input by catching exceptions.

```c#
namespace System.Globalization
{
    public sealed class IdnMapping
    {
        // Existing API
        public string GetAscii(string unicode);
        public string GetAscii(string unicode, int index);
        public string GetAscii(string unicode, int index, int count);

        public string GetUnicode(string ascii);
        public string GetUnicode(string ascii, int index);
        public string GetUnicode(string ascii, int index, int count);

        // Proposed API
        public string GetAscii(ReadOnlySpan<char> unicode);
        public bool TryGetAscii(ReadOnlySpan<char> unicode, out string ascii);
        public bool TryGetAscii(ReadOnlySpan<char> unicode, Span<char> destination, out int charsWritten);

        public string GetUnicode(ReadOnlySpan<char> ascii);
        public bool TryGetUnicode(ReadOnlySpan<char> ascii, out string unicode);
        public bool TryGetUnicode(ReadOnlySpan<char> ascii, Span<char> destination, out int charsWritten);
    }
}
```

The `TryGet*` methods would return false in the case of invalid input/insufficient space in the destination span.

This new API would simplify call sites and remove allocations throughout code dealing with internationalized domain names, like Uri and Markdig.

cc: @tarekgh

**Comment by tarekgh** [comment]
In general, the proposal looks good. 

Why we need the APIs:

```C#
    public string GetAscii(ReadOnlySpan<char> unicode);
    public string GetUnicode(ReadOnlySpan<char> ascii);
```

I don't think these are useful if we are going to allocate a string anyway. And the other proposed APIs can be used at that time. What do you think about that?

**Comment by MihaZupan** [comment]
Those aren't needed in my use-cases, as `Try*` would always be used.

I'll remove them as they could easily be exposed later if a use-case presents itself.

**Comment by stephentoub** [comment]
I don't understand these APIs:
```
public bool TryGetAscii(ReadOnlySpan<char> unicode, out string ascii);
public bool TryGetUnicode(ReadOnlySpan<char> ascii, out string unicode);
```
What does the `Try` mean here?  If it's to avoid throwing in the case where the data is somehow invalid, that's a different meaning than the other `Try` overloads would have, which would be based solely on whether the destination is large enough.  The Boolean returned from a `Try` is supposed to convey only one thing, and in such span-based Try methods, it's always used to connote whether the destination was large enough to store the transformed data.  I don't think we want two overloads of the same method having a different meaning for the `Try`.

**Comment by MihaZupan** [comment]
The idea was for both the `out string` and `Span destination` to return false on invalid input. The span one would also return false on insufficient space.

I agree the Span overload would be confusing to use since you couldn't differentiate between invalid input/insufficient space, without ensuring you supply a worst-case sized buffer.

Do we have a pattern of `Try*` methods ever throwing? If so, we could have all overloads throw on invalid input, where the `Try` only returns false on insufficient space.
```c#
bool GetAscii(ReadOnlySpan<char> unicode, out string ascii); // No Try
bool TryGetAscii(ReadOnlySpan<char> unicode, Span<char> destination, out int charsWritten);
```

Alternatively, we would need an `OperationStatus`-style return?

**Comment by stephentoub** [comment]
> Do we have a pattern of Try* methods ever throwing?

Yes, Try methods can still throw.

> Alternatively, we would need an OperationStatus-style return?

Why not just:
```C#
string GetAscii(ReadOnlySpan<char> unicode);
bool TryGetAscii(ReadOnlySpan<char> unicode, Span<char> destination, out int charsWritten);
```
?

**Comment by MihaZupan** [comment]
> `string GetAscii(ReadOnlySpan<char> unicode);`

Yes, that is the shape we'd want (lazy copy-pasting, I've edited the comment).

A few examples of exceptional inputs from reading IdnMapping code:
- Length: Empty / over 255 chars total / over 63 chars per label
- Certain sequences
  * Two dots in a row
  * (`-` before `.`): `foo-.bar`
  * Invalid right-to-left
- etc.
- There are a few more cases in [test code](https://github.com/dotnet/runtime/blob/acd8e5d4c32b42e0d0ec457e0347f1d9d1eefada/src/libraries/System.Globalization.Extensions/tests/IdnMapping/IdnMappingGetAsciiTests.cs).

I wouldn't expect the performance overhead of exceptions to be a problem for well-behaving apps.

**Comment by ericstj** [comment]
Moving this to 6.0.0 as it is not yet approved and we're driving 5.0.0 issues down.  Move it back if you think this is still needed in 5.0.0.

**Comment by ericstj** [comment]
This is too late for 6.0.0 at this point.

**Comment by terrajobst** [comment]
* Looks good as proposed

```C#
namespace System.Globalization;

public sealed class IdnMapping
{
    // Existing API
    // public string GetAscii(string unicode);
    // public string GetAscii(string unicode, int index);
    // public string GetAscii(string unicode, int index, int count);
    //
    // public string GetUnicode(string ascii);
    // public string GetUnicode(string ascii, int index);
    // public string GetUnicode(string ascii, int index, int count);

    // Proposed API
    public string GetAscii(ReadOnlySpan<char> unicode);
    public bool TryGetAscii(ReadOnlySpan<char> unicode, Span<char> destination, out int charsWritten);

    public string GetUnicode(ReadOnlySpan<char> ascii);
    public bool TryGetUnicode(ReadOnlySpan<char> ascii, Span<char> destination, out int charsWritten);
}
```

**Comment by stephentoub** [comment]
@MihaZupan, there are fast paths available for the current methods, where the original string is returned if no encoding is required. What is the intent for the overloads here that take a span and return a string? Won't those now force allocation in cases where the existing APIs didn't?

**Comment by MihaZupan** [comment]
Yes, they would possibly allocate more if the caller changed from `GetAscii(s, i, length)` and happened to have `0` and `s.Length` offsets.
The idea is similar as with e.g. `Uri.{Un}EscapeDataString(span)`, so you can use the simple helper with inputs other than string.

Looking through the usages in the BCL and Markdig, we don't have a need for it. Inputs are either already just `string`, or we'd also want to avoid the result allocation. I'd be fine with just leaving those out.
